README.md

```diff
@@ -1,13 +1,340 @@
+
+All dependencies are cached globally in `$HOME/.ambiata/mafia/packages`
+using a nix-like hashing system so that for a given set of transitive
+dependencies, a package is only ever built once. Source dependencies are
+also cached, using their dependencies and source code as the hash. This
+is critical for Haskell development at scale. At Ambiata we have around
+150 Haskell packages which are being developed at a rapid pace, so it's
+crucial that we don't need to think about curating blessed sets of
+packages that we can cache as a single snapshot.
+
+
+Commands
+--------
+
+Many of Mafia's commands are, more or less, direct wrappers of Cabal
+commands. The main difference is that they have a pre-step which takes
+care of installing any dependencies your project has, so you never have
+to think about it.
+
+### mafia build
+
+Builds a project, including all executables, test suites and benchmarks.
+
+Mafia uses whichever GHC it finds on the `PATH` to build a project. It
+keeps Cabal sandboxes in version specific directories, so that it is
+easy to [switch](https://gist.github.com/jystic/efded8238bc4c64affb7)
+between versions of GHC while developing a project.
+
+Using the switching setup linked above, a typical multi-GHC session
+might look as follows:
+
+```
+$ g
+The Glorious Glasgow Haskell Compilation System, version 7.8.4
+$ mafia build
+Writing a default package environment file to
+/Users/bob/src/foo/cabal.sandbox.config
+Creating a new sandbox at
+/Users/bob/src/foo/.cabal-sandbox/7.8.4
+Checking for changes to dependencies...
+Installing dependencies...
+-- snip --
+$ g
+The Glorious Glasgow Haskell Compilation System, version 7.10.2
+$ mafia build
+Writing a default package environment file to
+/Users/bob/src/foo/cabal.sandbox.config
+Creating a new sandbox at
+/Users/bob/src/foo/.cabal-sandbox/7.10.2
+Checking for changes to dependencies...
+Installing dependencies...
+-- snip --
+```
+
+Mafia always works in conjunction with cabal and tries to stay as
+compatible as possible. This means that after a `mafia build` it should
+be possible to run cabal commands (such as `cabal exec`) and have cabal
+benefit from the sandbox which Mafia has constructed.
+
+Mafia makes it easy to build your executable with profiling enabled,
+just run `mafia build -p`. Mafia will then build profiling versions of
+all the packages which your project depends on. Packages in the global
+cache are upgraded to support profiling in-place, so you don't need to
+perform the regular build for those packages all over again.
+
+
+### mafia clean
+
+Clean up after a build. This removes the Cabal sandbox and the `dist` directory.
+
+
+### mafia test
+
+Test this project, by default this runs all test suites.
+
+
+### mafia bench
+
+Benchmark this project, by default this runs all the benchmarks.
+
+
+### mafia repl
+
+Start GHCi via `cabal repl`, by default on the library of the project.
+
+
+### mafia quick
+
+Start GHCi directly on one or more files.
```

Comment:
Hmm, not really sure what to say, ideally i'd like to get rid of repl and have quick be the only option if we can make it work in all the cases